### PR TITLE
Don't block on forwarding request response, populate records that failed to forward in local CPF buffer

### PR DESCRIPTION
**Pull request by engechas**
### Description
This PR refactors the `RemotePeerForwarder` to avoid blocking on the `PeerForwarderClient`'s calls to forward records. This allows the local `ProcessWorker` to continue making progress while the network calls to a peer are performed in the background. If a background forwarding request fails, the records are written the local peer forwarding buffer for processing.

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java [hunk 1]
```diff
@@ -110,25 +124,37 @@ private WireEvent getWireEvent(final Event event) {
...
-            return aggregate.join();
-        }, executorService);
+            try {
+                final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
+                return aggregate.join();
+            } catch (final Exception e) {
+                processFailedRequestsLocally(null, records, peerForwarderReceiveBuffer);
+                throw e;
+            }
+        }, executorService).thenApply(httpResponse -> processFailedRequestsLocally(httpResponse, records, peerForwarderReceiveBuffer));
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:139]: I'd imagine it would be possible for the process worker to go faster than then peer requests and then these threads from the `executorService` fill up. Will the call block the processor thread when the `executorService`'s thread pool is exhausted?

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java [hunk 2]
```diff
@@ -110,25 +124,37 @@ private WireEvent getWireEvent(final Event event) {
...
-        try {
-            return aggregatedHttpResponseCompletableFuture.get();
-        } catch (final InterruptedException e) {
-            LOG.error("Peer forwarding interrupted.");
-            throw new RuntimeException(e);
-        } catch (final ExecutionException e) {
-            if (e.getCause() instanceof RuntimeException)
-                throw (RuntimeException) e.getCause();
-            throw new RuntimeException(e.getCause());
+    private AggregatedHttpResponse processFailedRequestsLocally(final AggregatedHttpResponse httpResponse, final Collection<Record<Event>> records,
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:142]: From a class-design perspective, I don't think the `PeerForwarderClient` should be responsible for the logic of handling failed requests.

I would suggest that you move `processFailedRequestsLocally` to the `RemotePeerForwarder` class. Then the `PeerForwarderClient` simply receives a method (perhaps a `Consumer<AggregatedHttpResponse>`). Then in `RemotePeerForwarder`:

```
 peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
                             destinationIp, pluginId, pipelineName, failedResponse -> processFailedRequestsLocally(failedResponse, entry.getValue(), peerForwarderReceiveBuffer);
```

You could even take the step of moving the `processFailedRequestsLocally` into a dedicated class for error handling. It might make unit testing easier.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 3]
```diff
@@ -31,16 +29,13 @@
 
 class RemotePeerForwarder implements PeerForwarder {
     private static final Logger LOG = LoggerFactory.getLogger(RemotePeerForwarder.class);
-    private static final int READ_BATCH_DELAY = 3_000;
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:34]: Why did you remove this line in this PR? Doesn't #2159 do this?

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java [hunk 4]
```diff
@@ -110,25 +124,37 @@ private WireEvent getWireEvent(final Event event) {
...
+                                                                final PeerForwarderReceiveBuffer<Record<Event>> peerForwarderReceiveBuffer) {
+        if (httpResponse == null || httpResponse.status() != HttpStatus.OK) {
+            try {
+                peerForwarderReceiveBuffer.writeAll(records, FAILED_REQUESTS_BUFFER_WRITE_TIMEOUT_MILLIS);
+            } catch (final Exception e) {
+                LOG.error("Unable to write failed records to local peer forwarder receive buffer due to exception. Dropping data.", e);
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:148]: It could be good to log how many records were dropped in this log line.

**Review by dlvenable — COMMENTED** [review]
Nice work!

FYI, You can create "draft" PRs in GitHub in the future to signal that something is a work-in-progress.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java (hunk 3 above)
Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:34]: I think removing this was a subtle miss from #2159. It wasn't removed in that PR even though it is no longer used
https://github.com/opensearch-project/data-prepper/pull/2159/files#diff-1cbb2edcb637c79fa75c0ee50aee7d939632e0fa53289bb31da9dbd823bd4e85R34

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java (hunk 4 above)
Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:148]: Good call, will add

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java (hunk 1 above)
Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:139]: That's a good question. Let me see if I can simulate this scenario. It probably makes sense to have a dedicated threadpool here regardless

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java (hunk 2 above)
Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:142]: Fair point. Passing the buffer to the client felt wrong, I like this approach better

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:139]: I agree that we should have a dedicated thread-pool.

I am wanting to be sure that we handle the situation where the thread pool is exhausted correctly. If this throws an exception that would be bad. I think blocking to wait is acceptable in my opinion. And I'd be concerned about having some secret buffer of queued requests that we have no real insight into.

**Comment by engechas** [comment]
This is ready for another look @dlvenable

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java [hunk 5]
```diff
@@ -78,7 +78,7 @@ public <T> T gauge(String name, T obj, ToDoubleFunction<T> valueFunction) {
         return Metrics.gauge(getMeterName(name), obj, valueFunction);
     }
 
-    private String getMeterName(final String name) {
+    public String getMeterName(final String name) {
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java:81]: I don't see this being used here. Do we still need to expose this method?

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 6]
```diff
@@ -66,20 +69,25 @@ class RemotePeerForwarder implements PeerForwarder {
...
+        this.failedForwardingRequestLocalWriteTimeout = failedForwardingRequestLocalWriteTimeout;
+        this.forwardingRequestExecutor = forwardingRequestExecutor;
         recordsActuallyProcessedLocallyCounter = pluginMetrics.counter(RECORDS_ACTUALLY_PROCESSED_LOCALLY);
         recordsToBeProcessedLocallyCounter = pluginMetrics.counter(RECORDS_TO_BE_PROCESSED_LOCALLY);
         recordsToBeForwardedCounter = pluginMetrics.counter(RECORDS_TO_BE_FORWARDED);
         recordsSuccessfullyForwardedCounter = pluginMetrics.counter(RECORDS_SUCCESSFULLY_FORWARDED);
         recordsFailedForwardingCounter = pluginMetrics.counter(RECORDS_FAILED_FORWARDING);
-        recordsReceivedFromPeersCounter = pluginMetrics.counter(RECORDS_RECEIVED_FROM_PEERS);
+        recordsReceivedFromPeersCount = new AtomicInteger(0);
+        pluginMetrics.gauge(RECORDS_RECEIVED_FROM_PEERS, recordsReceivedFromPeersCount);
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:90]: It appears that you made this a gauge so that you could decrease for all the records moved locally. Do you think there might a be a different way to avoid counting these while still using a counter?

This is technically a breaking change because are changing the metric (including the name Micrometer produces).

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 7]
```diff
@@ -98,22 +106,20 @@ public Collection<Record<Event>> forwardRecords(final Collection<Record<Event>>
                 recordsToBeProcessedLocallyCounter.increment(entry.getValue().size());
             } else {
                 recordsToBeForwardedCounter.increment(entry.getValue().size());
-                AggregatedHttpResponse httpResponse;
                 try {
-                    httpResponse = peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
-                            destinationIp, pluginId, pipelineName);
+                    forwardingRequestExecutor.submit(() -> {
+                        try {
+                            peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(), destinationIp, pluginId,
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:112]: Looking at this code, we should probably split serializing from sending. It could help with the error handling. But, that is beyond the scope of this PR.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java (hunk 6 above)
Comment (dlvenable) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:90]: Perhaps we can move the existing counter to `PeerForwarderHttpService` instead of having it here.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java (hunk 5 above)
Comment (engechas) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java:81]: It is used in the tests. Some of the other tests were using local string concatenation to get this value. I figured there wasn't any harm in exposing this method since it's read-only and allows for a single source of truth in the formatting.

Let me know if you are strongly against exposing this method and I'll swap to local string concatenation.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java (hunk 6 above)
Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:90]: I did not realize this was breaking. I'll look into moving it to the HttpService

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java (hunk 5 above)
Comment (dlvenable) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java:81]: I'm not strongly opposed. I didn't see it used at first. But, it is because some diffs were hidden by GitHub. If you need it, I'm fine with adding it.

Comment (engechas) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/PluginMetrics.java]: I was to able to get rid of this change by moving the existing counter

**Review by dlvenable — DISMISSED** [review]
Looks great! Thanks!

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 8]
```diff
@@ -98,22 +101,13 @@ public Collection<Record<Event>> forwardRecords(final Collection<Record<Event>>
...
-                    httpResponse = peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
-                            destinationIp, pluginId, pipelineName);
+                    submitForwardingRequest(entry.getValue(), destinationIp);
                 } catch (final Exception ex) {
-                    httpResponse = null;
                     LOG.warn("Unable to send request to peer, processing locally.", ex);
-                }
-
-                if (httpResponse == null || httpResponse.status() != HttpStatus.OK) {
                     recordsToProcessLocally.addAll(entry.getValue());
```
Comment (asifsmohammed) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:108]: This shouldn't happen here. You are adding the failed records to buffer on line 197 and also returning them which makes the failed records to be processed twice.

Since we are making async call to forward records, we should return an empty collection here and let the buffer handle failed records

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 9]
```diff
@@ -98,22 +101,13 @@ public Collection<Record<Event>> forwardRecords(final Collection<Record<Event>>
...
             } else {
                 recordsToBeForwardedCounter.increment(entry.getValue().size());
-                AggregatedHttpResponse httpResponse;
                 try {
-                    httpResponse = peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
-                            destinationIp, pluginId, pipelineName);
+                    submitForwardingRequest(entry.getValue(), destinationIp);
                 } catch (final Exception ex) {
-                    httpResponse = null;
                     LOG.warn("Unable to send request to peer, processing locally.", ex);
```
Comment (asifsmohammed) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:107]: I think this LOG is redundant, we can either remove it from here or on line 186

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java [hunk 10]
```diff
@@ -72,12 +69,12 @@ public AggregatedHttpResponse serializeRecordsAndSendHttpRequest(
```
Comment (asifsmohammed) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java:69]: `ExecutorService` created in this class is not being used anymore, can you remove that?

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java [hunk 11]
```diff
@@ -87,7 +89,9 @@ public PeerForwarderConfiguration (
             @JsonProperty("batch_size") final Integer batchSize,
             @JsonProperty("batch_delay") final Integer batchDelay,
             @JsonProperty("buffer_size") final Integer bufferSize,
-            @JsonProperty("drain_timeout") final Duration drainTimeout
+            @JsonProperty("drain_timeout") final Duration drainTimeout,
+            @JsonProperty("failed_forwarding_requests_local_write_timeout") final Integer failedForwardingRequestLocalWriteTimeout,
+            @JsonProperty("forwarding_requests_workers") final Integer forwardingRequestWorkers
```
Comment (asifsmohammed) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java:94]: This option is similar to `client_thread_count` which is deprecated with this change. Can you remove that as well?

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 12]
```diff
@@ -130,7 +124,6 @@ public Collection<Record<Event>> receiveRecords() {
         // Checkpoint the current batch read from the buffer after reading from buffer
         peerForwarderReceiveBuffer.checkpoint(checkpointState);
 
-        recordsReceivedFromPeersCounter.increment(records.size());
```
Comment (asifsmohammed) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:133]: Why is this metric moved from here to `PeerForwarderHttpService`?

**Review by asifsmohammed — DISMISSED** [review]

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java (hunk 11 above)
Comment (dlvenable) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java]: What was `client_thread_count` used for beforehand?

Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java]: It was used for the ExecutorService in the PeerForwarderClient: https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java#L55

Comment (dlvenable) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderConfiguration.java]: OK. So it really is the same thing from what I can see.

**Review by dlvenable — DISMISSED** [review]

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java [hunk 13]
```diff
@@ -98,22 +101,13 @@ public Collection<Record<Event>> forwardRecords(final Collection<Record<Event>>
...
-                    httpResponse = peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(),
-                            destinationIp, pluginId, pipelineName);
+                    submitForwardingRequest(entry.getValue(), destinationIp);
                 } catch (final Exception ex) {
-                    httpResponse = null;
-                    LOG.warn("Unable to send request to peer, processing locally.", ex);
-                }
-
-                if (httpResponse == null || httpResponse.status() != HttpStatus.OK) {
+                    LOG.warn("Unable submit request for forwarding, processing locally.", ex);
```
[the remark this answers is not in the document]
Comment (engechas) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java:107]: Oops, good catch

**Review by asifsmohammed — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
